### ISSUE TITLE
Imaging mass cytometry data analysis

### ISSUE THREAD

**Issue by StiThor** [BodenmillerGroup/IMCDataAnalysis#59]
Hi Nils! First of all thanks for providing IMC users with great analysis tools :) I`m trying to use the pipeline you have made and everything works fine until part 9.3. This is done using the example dataset provided in the pipeline. I have put hashtags around the errors occurring running the script as bellow: 

 spe <- readRDS("C:/Bodenmiller/spe.rds")
> library(dplyr)
> cluster_celltype <- recode(spe$nn_clusters_corrected,
+                            "1" = "Tumor_proliferating",
+                            "2" = "Neutrophil",
+                            "3" = "Tumor",
+                            "4" = "Tumor",
+                            "5" = "Stroma",
+                            "6" = "Proliferating",
+                            "7" = "Myeloid",
+                            "8" = "CD8",
+                            "9" = "Plasma_cell",
+                            "10" = "CD4",
+                            "11" = "Bcell",
+                            "12" = "Stroma",
+                            "13" = "pDC")
> spe$cluster_celltype <- cluster_celltype
> library(cytomapper)
> if (interactive()) {
+   
+   images <- readRDS("C:/Bodenmiller/images.rds")
+   masks <- readRDS("C:/Bodenmiller/masks.rds")
+   
+   cytomapperShiny(object = spe, mask = masks, image = images, 
+                   cell_id = "ObjectNumber", img_id = "sample_id")
+ }
The `name` provided ('') does not correspond to a known icon
The `name` provided ('') does not correspond to a known icon
The `name` provided ('') does not correspond to a known icon
The `name` provided ('') does not correspond to a known icon
Loading required package: shiny

Listening on http://127.0.0.1:6398

> celltype <- setNames(c("#3F1B03", "#F4AD31", "#894F36", "#1C750C", "#EF8ECC", 
+                        "#6471E2", "#4DB23B", "grey", "#F4800C", "#BF0A3D", "#066970"),
+                      c("Tumor", "Stroma", "Myeloid", "CD8", "Plasma_cell", 
+                        "Treg", "CD4", "undefined", "BnTcell", "Bcell", "Neutrophil"))
> metadata(spe)$color_vectors$celltype <- celltype
> library(SingleCellExperiment)
> label_files <- list.files("C:/Bodenmiller/gated_cells", 
+                           full.names = TRUE, pattern = ".rds$")
> # Read in SPE objects
> spes <- lapply(label_files, readRDS)
> # Merge SPE objects
> concat_spe <- do.call("cbind", spes)
#############################################################################
**'sample_id's are duplicated across 'SpatialExperiment' objects to cbind; appending sample indices.**
#############################################################################
> filter_labels <- function(object, 

+                           label = "cytomapper_CellLabel") {
+   cur_tab <- unclass(table(colnames(object), object[[label]]))
+   
+   cur_labels <- colnames(cur_tab)[apply(cur_tab, 1, which.max)]
+   names(cur_labels) <- rownames(cur_tab)
+   
+   cur_labels <- cur_labels[rowSums(cur_tab) == 1]
+   
+   return(cur_labels)
+ }
> labels <- filter_labels(concat_spe)
> cur_spe <- concat_spe[,concat_spe$cytomapper_CellLabel != "Tumor"]
################################################################
Error in validObject(result) : invalid class “SpatialExperiment” object: 
    no 'spatialData' field in 'int_colData'
################################################################
> non_tumor_labels <- filter_labels(cur_spe)
#################################################################
 Error in h(simpleError(msg, call)) : 
error in evaluating the argument 'x' in selecting a method for function 'colnames': object 'cur_spe' not found
################################################################

**Comment by nilseling** [comment]
Hi @StiThor 

this is a bit tough to debug as I can't reproduce the issue on my side. Could you please make sure that you downloaded the correct data (`gated_cells`) and that you are working with the newest package versions? The warning that sample_ids are duplicated is to be expected. For the other errors I have no clue where they are coming from. It could also help to start with a fresh R session and test it again.

**Comment by StiThor** [comment]
Thanks for the quick response :) The "gate_cells" I downloaded is this one:
https://zenodo.org/record/7647079/files/gated_cells.zip

Is that correct?

**Comment by nilseling** [comment]
That is correct. What's the output of `sessionInfo()`?

**Comment by StiThor** [comment]
R version 4.1.3 (2022-03-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19045)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] shiny_1.7.4                 dplyr_1.1.0                 cytomapper_1.11.0          
 [4] SingleCellExperiment_1.16.0 SummarizedExperiment_1.24.0 Biobase_2.54.0             
 [7] GenomicRanges_1.46.1        GenomeInfoDb_1.30.1         IRanges_2.28.0             
[10] S4Vectors_0.32.4            BiocGenerics_0.40.0         MatrixGenerics_1.6.0       
[13] matrixStats_0.63.0          EBImage_4.36.0             

loaded via a namespace (and not attached):
  [1] utf8_1.2.3                  shinydashboard_0.7.2        R.utils_2.12.2             
  [4] tidyselect_1.2.0            htmlwidgets_1.6.2           grid_4.1.3                 
  [7] BiocParallel_1.28.3         Rtsne_0.16                  DropletUtils_1.14.2        
 [10] aws.signature_0.6.0         flowCore_2.6.0              ScaledMatrix_1.2.0         
 [13] munsell_0.5.0               ragg_1.2.5                  codetools_0.2-18           
 [16] interp_1.1-3                withr_2.5.0                 colorspace_2.1-0           
 [19] rstudioapi_0.14             ggsignif_0.6.4              GenomeInfoDbData_1.2.7     
 [22] polyclip_1.10-4             farver_2.1.1                pheatmap_1.0.12            
 [25] rhdf5_2.38.1                flowWorkspace_4.6.0         TH.data_1.1-1              
 [28] vctrs_0.6.0                 generics_0.1.3              R6_2.5.1                   
 [31] doParallel_1.0.17           ggbeeswarm_0.7.1            clue_0.3-64                
 [34] rsvd_1.0.5                  locfit_1.5-9.6              cachem_1.0.7               
 [37] bitops_1.0-7                rhdf5filters_1.6.0          DelayedArray_0.20.0        
 [40] promises_1.2.0.1            scales_1.2.1                multcomp_1.4-23            
 [43] beeswarm_0.4.0              gtable_0.3.3                beachmat_2.10.0            
 [46] RProtoBufLib_2.6.0          sandwich_3.0-2              rlang_1.1.0                
 [49] systemfonts_1.0.4           GlobalOptions_0.1.2         splines_4.1.3              
 [52] rstatix_0.7.2               hexbin_1.28.3               broom_1.0.4                
 [55] reshape2_1.4.4              yaml_2.3.7                  abind_1.4-5                
 [58] backports_1.4.1             httpuv_1.6.9                RBGL_1.70.0                
 [61] tools_4.1.3                 SpatialExperiment_1.4.0     ggplot2_3.4.1              
 [64] ellipsis_0.3.2              jquerylib_0.1.4             raster_3.6-20              
 [67] RColorBrewer_1.1-3          ggridges_0.5.4              Rcpp_1.0.10                
 [70] plyr_1.8.8                  base64enc_0.1-3             sparseMatrixStats_1.6.0    
 [73] zlibbioc_1.40.0             purrr_1.0.1                 RCurl_1.98-1.9             
 [76] FlowSOM_2.2.0               ggpubr_0.6.0                deldir_1.0-6               
 [79] GetoptLong_1.0.5            viridis_0.6.2               cowplot_1.1.1              
 [82] fontawesome_0.5.0           zoo_1.8-11                  CATALYST_1.18.1            
 [85] ggrepel_0.9.3               cluster_2.1.2               colorRamps_2.3.1           
 [88] magrittr_2.0.3              ncdfFlow_2.40.0             data.table_1.14.8          
 [91] magick_2.7.3                scattermore_0.8             circlize_0.4.15            
 [94] mvtnorm_1.1-3               ggnewscale_0.4.8            mime_0.12                  
 [97] fftwtools_0.9-11            xtable_1.8-4                XML_3.99-0.14              
[100] jpeg_0.1-10                 gridExtra_2.3               shape_1.4.6                
[103] ggcyto_1.22.0               scater_1.22.0               compiler_4.1.3             
[106] tibble_3.2.1                crayon_1.5.2                ggpointdensity_0.1.0       
[109] R.oo_1.25.0                 htmltools_0.5.4             later_1.3.0                
[112] tiff_0.1-11                 tidyr_1.3.0                 RcppParallel_5.1.7         
[115] aws.s3_0.3.21               DBI_1.1.3                   tweenr_2.0.2               
[118] ComplexHeatmap_2.10.0       MASS_7.3-55                 Matrix_1.5-3               
[121] car_3.1-2                   cli_3.6.0                   R.methodsS3_1.8.2          
[124] parallel_4.1.3              igraph_1.4.1                pkgconfig_2.0.3            
[127] sp_1.6-0                    terra_1.7-18                scuttle_1.4.0              
[130] xml2_1.3.3                  foreach_1.5.2               svglite_2.1.1              
[133] bslib_0.4.2                 vipor_0.4.5                 dqrng_0.3.0                
[136] XVector_0.34.0              drc_3.0-1                   stringr_1.5.0              
[139] digest_0.6.31               ConsensusClusterPlus_1.58.0 graph_1.72.0               
[142] edgeR_3.36.0                DelayedMatrixStats_1.16.0   curl_5.0.0                 
[145] gtools_3.9.4                rjson_0.2.21                lifecycle_1.0.3            
[148] jsonlite_1.8.4              Rhdf5lib_1.16.0             BiocNeighbors_1.12.0       
[151] carData_3.0-5               viridisLite_0.4.1           limma_3.50.3               
[154] fansi_1.0.4                 pillar_1.9.0                lattice_0.20-45            
[157] plotrix_3.8-2               fastmap_1.1.1               httr_1.4.5                 
[160] survival_3.5-5              glue_1.6.2                  png_0.1-8                  
[163] iterators_1.0.14            svgPanZoom_0.3.4            Rgraphviz_2.38.0           
[166] sass_0.4.5                  stringi_1.7.12              ggforce_0.4.1              
[169] HDF5Array_1.22.1            nnls_1.4                    textshaping_0.3.6          
[172] BiocSingular_1.10.0         CytoML_2.6.0                memoise_2.0.1              
[175] latticeExtra_0.6-30         cytolib_2.6.2               irlba_2.3.5.1

**Comment by nilseling** [comment]
Hmm, this is quite old. Could you please update to R4.3 and Bioconductor 3.17 and test again?

**Comment by nilseling** [comment]
You could also test to explicitly load `SpatialExperiment` via `library(SpatialExperiment)`

**Comment by StiThor** [comment]
Thanks so much, Nils! :) It was just to update everything and then it ran smoothly! Have a great weekend!

**Comment by nilseling** [comment]
Yeah, thinking about it, the gated cells objects were generated by a never version of SpatialExperiment than the one you were using. That causes issues. Closing now

**Comment by StiThor** [comment]
That makes sense! Very eager to try this out now. I have around 400 tissue samples to analyze and this was the last problem to run the whole thing. The only challenge now is the ram of the computer to run all samples ;) Do you have any idea how much I will need to analyze around 7 million cells? I was thinking 128 gb ram

**Comment by nilseling** [comment]
Probably even more. I guess you would need to process and analyse your data in batches. You might also want to look for additional analysis approaches that performs analysis on disk.

**Comment by jwindhager** [comment]
Glad this could be resolved, thanks @nilseling  :-)

@StiThor, regarding the large dataset size, just as a sidenote (doesn't help much for the actual analysis): For many commands, steinbock supports on-disk (i.e., memory-mapped) processing of data; have a look at the `--help` of the individual commands, this will often offer an `--mmap` option (e.g. for the measurement commands).

**Comment by StiThor** [comment]
Do you guys have experience analyzing this amount of data with the pipeline you made? Jonas, when you say Steinbock do you mean the segmentation part? Or do you mean to run the analysis R code with Steinbock?

**Comment by jwindhager** [comment]
Just referring to the processing (data preprocessing, image segmentation, feature extraction) part with my OT comment, sorry for being unclear

**Comment by StiThor** [comment]
Oh yeah, that part actually went quite nicely using a laptop with only 8 GB ram. I used 10 markers for the segmentation and it went on for like 3 hours. I posted a figure with the segmentation control in your script on LinkedIn. If you want to check it out go here :)
https://www.linkedin.com/posts/stian-tornaas-45a189165_segmentation-of-cells-is-the-foundation-for-activity-7046058510970699776-jYeu?utm_source=share&utm_medium=member_desktop
